lib/rust/ast/generation/src/api.rs

```diff
@@ -0,0 +1,672 @@
+        Module {name:self.name.clone(), lines:&self.content.items[..]}.write(source);
+    }
+}
+
+impl<'a> Generator<Collector> for &Module<'a> {
+    fn write(self, source:&mut Collector) {
+        source.module.push(self.name.clone());
+        for item in self.lines {
+            match item {
+                syn::Item::Mod   (val) => Module::from(val).write(source),
+                syn::Item::Type  (val) => TypeAlias::from(val).write(source),
+                syn::Item::Struct(val) => Class::from(val).write(source),
+                syn::Item::Enum  (val) => Enum::from(val).write(source),
+                _                      => (),
+            }
+        }
+        source.module.pop();
+    }
+}
+
+impl Generator<Collector> for &TypeAlias {
+    fn write(self, source:&mut Collector) {
+        if self.val.args.is_empty() {
+            source.monomorphize(&self.val);
+        }
+    }
+}
+
+impl Generator<Collector> for &Class {
+    fn write(self, source:&mut Collector) {
+        source.add_class(self.clone(), None);
+        source.types.insert(self.typ.name.clone());
+        if self.typ.args.is_empty() {
+            source.generics.entry(self.typ.name.clone()).or_default().insert(self.typ.clone());
+        }
+    }
+}
+
+impl Generator<Collector> for &Enum {
+    fn write(self, source:&mut Collector) {
+        for (_, variant) in self.variants.iter().cloned() {
+            source.add_class(variant, Some(self.typ.name.clone()));
+        }
+        source.types.insert(self.typ.name.clone());
+        if self.typ.args.is_empty() {
+            source.generics.entry(self.typ.name.clone()).or_default().insert(self.typ.clone());
+        }
+    }
+}
+
+
+// === ToTokens Impls ===
+
+impl ToTokens for Name {
+    fn to_tokens(&self, tokens:&mut TokenStream) {
+        syn::Ident::new(&self.str, proc_macro2::Span::call_site()).to_tokens(tokens)
+    }
+}
+
+
+
+// ======================
+// === Rust Generator ===
+// ======================
+
+/// An associated types in trait with monorphized signature.
+#[derive(Debug,Clone)]
+pub struct AssociatedType {
+    /// Name of the associated type.
+    pub name: Name,
+    /// Name of the enum variant if any.
+    pub variant: Option<Name>,
+    /// The class the associated type represents.
+    pub class: Class
+}
+
+impl AssociatedType {
+    /// Ast tree of any type used in trait.
+    ///
+    /// For custom types this returns `<Self as Api>::Name`.
+    /// For builtin types this returns `Name<typ(arg1), typ(arg2)..>`.
+    pub fn typ(typ:&Type) -> TokenStream {
+        let args = typ.args.iter().map(Self::typ);
+        let name = &typ.name;
+        if types::builtin(&name).is_none() {
+            quote!(<Self as Api>::#name)
+        } else {
+            quote!(#name<#(#args),*>)
+        }
+    }
+
+    /// An api of function that constructs the given associated type.
+    ///
+    /// For example `fn name(x:i64, y:<Self as Api>::Y) -> <Self as Api>::Name`
```

Comment:
This doumentation comment doesn't make any sense. Additionally, if you provide an example you should do it as a proper doctest example in rust.